spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java

```diff
@@ -364,187 +239,104 @@ private void rethrowAsLockException(Exception e) {
-				return true;
-			}
-
-			toHardThreadStorage(this);
-
-			/*
-			 * Set these now so they will be persisted if successful.
-			 */
-			this.lockedAt = System.currentTimeMillis();
-			this.threadName = currentThread.getName();
-
-			Boolean success = false;
 			try {
-
-				success = RedisLockRegistry.this.redisTemplate.execute(new RedisCallback<Boolean>() {
-
-					@Override
-					public Boolean doInRedis(RedisConnection connection) throws DataAccessException {
-
-						/*
-						Perform Redis command 'SET resource-name anystring NX EX max-lock-time' directly.
-						As it is recommended by Redis: http://redis.io/commands/set.
-						This command isn't supported directly by RedisTemplate.
-						*/
-						long expireAfter = TimeoutUtils.toSeconds(RedisLockRegistry.this.expireAfter,
-								TimeUnit.MILLISECONDS);
-						RedisSerializer<String> serializer = RedisLockRegistry.this.redisTemplate.getStringSerializer();
-						byte[][] actualArgs = new byte[][] {
-								serializer.serialize(constructLockKey()),
-								RedisLockRegistry.this.lockSerializer.serialize(RedisLock.this),
-								serializer.serialize("NX"),
-								serializer.serialize("EX"),
-								serializer.serialize(String.valueOf(expireAfter))
-						};
-
-						return connection.execute("SET", actualArgs) != null;
-					}
-
-				});
+				return tryLock(0, TimeUnit.MILLISECONDS);
 			}
-			finally {
-
-				if (!success) {
-					this.lockedAt = 0;
-					this.threadName = null;
-					toWeakThreadStorage(this);
-				}
-				else {
-					this.thread = currentThread;
-					if (logger.isDebugEnabled()) {
-						logger.debug("New lock; " + this);
-					}
-				}
-
+			catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
+				return false;
 			}
-
-			return success;
 		}
 
 		@Override
 		public boolean tryLock(long time, TimeUnit unit) throws InterruptedException {
-			Lock localLock = RedisLockRegistry.this.localRegistry.obtain(this.lockKey);
-			if (!localLock.tryLock(time, unit)) {
+			long now = System.currentTimeMillis();
+			if (!this.localLock.tryLock(time, unit)) {
 				return false;
 			}
 			try {
-				long expire = System.currentTimeMillis() + TimeUnit.MILLISECONDS.convert(time, unit);
+				long expire = now + TimeUnit.MILLISECONDS.convert(time, unit);
 				boolean acquired;
 				while (!(acquired = obtainLock()) && System.currentTimeMillis() < expire) { //NOSONAR
 					Thread.sleep(100); //NOSONAR
 				}
 				if (!acquired) {
-					localLock.unlock();
+					this.localLock.unlock();
 				}
 				return acquired;
 			}
 			catch (Exception e) {
-				localLock.unlock();
+				this.localLock.unlock();
 				rethrowAsLockException(e);
 			}
 			return false;
 		}
 
+		private boolean obtainLock() {
+			boolean success = RedisLockRegistry.this.redisTemplate.execute(RedisLockRegistry.this.obtainLockScript,
+					Collections.singletonList(this.lockKey), RedisLockRegistry.this.clientId,
+					String.valueOf(RedisLockRegistry.this.expireAfter));
```

Comment:
Now that you mention it I also think we're good with a simple `SET` command. I guess I started with `HMSET` since I expected we'll be storing more attributes and never revisited that.